composer/datasets/hf_streaming.py

```diff
@@ -0,0 +1,246 @@
+
+        devices_per_shard = 1
+        if world_size > num_shards:
+            log.warning(
+                f"Not enough unique shards ({num_shards}) for world size ({world_size}). Splitting shards among devices."
+            )
+            assert world_size % num_shards == 0, f"Cannot evenly split shards among devices"
+            devices_per_shard = world_size // num_shards
+        shard_offset = rank // devices_per_shard
+        device_offset = rank % devices_per_shard
+
+        device_filepaths = filepaths[shard_offset::world_size]
+        dataset._ex_iterable.kwargs['filepaths'] = device_filepaths
+
+        # Subsample dataset if shard is being shared among devices
+        # NOTE: Mapping is executed in batched mode for better CPU utilization,
+        # but the returned dataset is still an iterable over text samples
+        if devices_per_shard > 1:
+            dataset = dataset.map(
+                partial(self._subsample, device_offset),
+                batched=True,
+                batch_size=devices_per_shard,
+            )
+        return dataset
+
+    def _tokenize(self, text_batch):
+        # Convert a text batch to a token batch
+        if self.group_method == "truncate":
+            truncation = True
+            padding = 'max_length'
+            max_length = self.max_seq_len
+        else:
+            truncation = False
+            padding = False
+            max_length = None
+        return self.tokenizer(text_batch["text"], truncation=truncation, padding=padding, max_length=max_length)
+
+    def _group_tokens(self, token_batch):
+        if self.group_method == "concat":
+            # Concatenate all tokens.
+            concat_tokens = {}
+            num_tokens = None
+            for k, v in token_batch.items():
+                concat_v = list(chain(*v))
+                concat_tokens[k] = concat_v
+                if num_tokens is None:
+                    num_tokens = len(concat_v)
+                else:
+                    assert num_tokens == len(concat_v), "Not all values in concat_tokens dict have same len()"
+            assert num_tokens is not None, "Failed to determine num_tokens."
+
+            # We drop the small remainder of tokens at the end of the batch,
+            # In the future we could support padding.
+            if num_tokens >= self.max_seq_len:
+                num_tokens = (num_tokens // self.max_seq_len) * self.max_seq_len
+
+            # Split into token samples of size max_seq_len.
+            result = {
+                k: [v[i:i + self.max_seq_len] for i in range(0, num_tokens, self.max_seq_len)
+                   ] for k, v in concat_tokens.items()
+            }
+            result["labels"] = result["input_ids"].copy()
+            return result
+        else:
+            raise ValueError(f"Unknown group_method: '{self.group_method}'")
+
+    def initialize_object(self, batch_size: int, dataloader_hparams: DataloaderHparams) -> DataSpec:
```

Comment:
is there a way we can move all of this logic into a separate class? 

the motivation being that external users not using YAHP should be able to create their StreamingDataset and pass it to our trainer directly?

Reply:
Yes, I could definitely move some of the first few steps like sharding, and the overwritten __len__ into a class, but if users want to build their own custom dataloader, they would still have to build the collate functions and pass those correctly to their dataloader right? I'm also wondering if the preprocessing steps like (shuffle, tokenize, group) would belong in the class, or if again it is user-specific. I guess I'm skeptical how much I can abstract away, haha.

Looking at Vision, I think there are classes like `ImageFolder(..) -> Dataset` that you can start from, but building a dataloader is still pretty task-specific.

For users that want to build their own streaming dataloaders, I think the upcoming Vision Webdatasets PR would be the better example to follow, and that PR comes with helper classes too. 

WDYT? I'm happy to take a pass at making a light HFStreamingDataset class, or punt it to a future PR?

Reply:
Update: we discussed offline, and will refactor a standalone dataset class for say notebook or script users.